java/ql/src/utils/Stubs.qll

```diff
@@ -0,0 +1,474 @@
+            interface = concat(stubTypeName(getAnInterestingBaseType().(Interface)), ", ")
+          ) else (
+            int_kw = "" and interface = ""
+          )
+        )
+      )
+    else result = ""
+  }
+
+  language[monotonicAggregates]
+  private string stubMembers() {
+    result =
+      stubEnumConstants(this) + stubFakeConstructor(this) +
+        concat(Member m | m = this.getAGeneratedMember() | stubMember(m))
+  }
+
+  private Member getAGeneratedMember() {
+    (
+      result.getDeclaringType() = this
+      or
+      exists(NestedType nt | result = nt |
+        nt = nt.getSourceDeclaration() and
+        nt.getEnclosingType().getSourceDeclaration() = this
+      )
+    ) and
+    not result.isPrivate() and
+    not result instanceof StaticInitializer and
+    not result instanceof InstanceInitializer
+  }
+
+  final Type getAGeneratedType() {
+    result = getAnInterestingBaseType()
+    or
+    result = getAGeneratedMember().(Callable).getReturnType()
+    or
+    result = getAGeneratedMember().(Callable).getAParameter().getType()
+    or
+    result = getAGeneratedMember().(Field).getType()
+    or
+    result = getAGeneratedMember().(NestedType)
+  }
+}
+
+/**
+ * A declaration that should be generated.
+ * This is extended in client code to identify the actual
+ * declarations that should be generated.
+ */
+abstract class GeneratedDeclaration extends Element { }
+
+private class IndirectType extends GeneratedType {
+  IndirectType() {
+    this.getASubtype() instanceof GeneratedType
+    or
+    this.(GenericType).getAParameterizedType() instanceof GeneratedType
+    or
+    exists(GeneratedType t |
+      this = getAContainedType(t.getAGeneratedType()).(RefType).getSourceDeclaration()
+    )
+    or
+    this.getSourceDeclaration() instanceof GeneratedType
+    or
+    this = any(GeneratedType t).getSourceDeclaration()
+    or
+    exists(GeneratedType t | this = t.(BoundedType).getATypeBound().getType())
+    or
+    exists(GeneratedDeclaration decl |
+      decl.(Member).getDeclaringType().getSourceDeclaration() = this
+    )
+    or
+    this.(NestedType).getEnclosingType() instanceof GeneratedType
+    or
+    exists(NestedType nt | nt instanceof GeneratedType and this = nt.getEnclosingType())
+  }
+}
+
+private class RootGeneratedType extends GeneratedType {
+  RootGeneratedType() { this = any(GeneratedDeclaration decl).(RefType).getSourceDeclaration() }
+}
+
+private Type getAContainedType(Type t) {
+  result = t
+  or
+  result = getAContainedType(t.(ParameterizedType).getATypeArgument())
+}
+
+/**
+ * Specify packages to exclude.
+ * Do not generate any types from these packages.
+ */
+abstract class ExcludedPackage extends Package { }
+
+/** Exclude types from the standard library. */
+private class DefaultLibs extends ExcludedPackage {
+  DefaultLibs() { this.getName().matches(["java.%", "javax.%", "jdk.%", "sun.%"]) }
```

Comment:
It seems like we need stubs for `javax` packages? (At least `javax.servlet` seemed to need stubbing).

```suggestion
  DefaultLibs() { this.getName().matches(["java.%", "jdk.%", "sun.%"]) }
```